dotenv/src/iter.rs

```diff
@@ -31,14 +33,70 @@ impl<R: Read> Iter<R> {
     }
 }
 
+struct QuotedLines<B> {
+    buf: B,
+}
+
+fn is_complete(buf: &String) -> bool {
+    let mut escape = false;
+    let mut strong_quote = false;
+    let mut weak_quote = false;
+    let mut count = 0_u32;
+
+    for c in buf.chars() {
+        if escape {
+            escape = false
+        } else {
+            match c {
+                '\\' => escape = true,
+                '"' if !strong_quote => {
+                    count += 1;
+                    weak_quote = true
+                }
+                '\'' if !weak_quote => {
+                    count += 1;
+                    strong_quote = true
+                }
+                _ => (),
+            }
+        }
+    }
+    count % 2 == 0
+}
+
+impl<B: BufRead> Iterator for QuotedLines<B> {
+    type Item = Result<String>;
+
+    fn next(&mut self) -> Option<Result<String>> {
+        let mut buf = String::new();
+        loop {
+            match self.buf.read_line(&mut buf) {
+                Ok(0) => return None,
+                Ok(_n) => {
+                    if is_complete(&buf) {
```

Comment:
This looks to me like it would cause an exponential slowdown if a quote contained a large number of lines - it would have to loop through every character read so far for each new line.

Would it be possible to make this a stateful parser which keeps track of what quotes have been seen so far, and only looks at each character once be possible without making it significantly more complicated?

I mean, the current code won't slow down any current use cases, and env variables are unlikely to be very large. But it seems non-ideal to have an exponential slowdown if it's avoidable?

I'm asking as an observer: I don't have any ownership over dotenv.

Reply:
I'll have a go at it

Reply:
have it working:
https://github.com/dotenv-rs/dotenv/pull/78